mathics/core/symbols.py

```diff
@@ -72,6 +72,49 @@ def strip_context(name) -> str:
     return name
 
 
+# FIXME: move to new module element.py
+class NumericOperators:
```

Comment:
I like this, as a first step. However, notice that ``Plus[3,"a"]`` or ``Times[x,Graphics[{}]]``  are  perfectly valid ``Expression`` s in WL.

Reply:
Nevertheless, from the Python side, maybe it also makes sense to define specialized behaviors for expressions like
```(Python)
Integer(1)+Integer(2)
```
or 
```(Python)
Integer(3)+MachineReal(.12)
```

Reply:
The issue here is more about what we need to express in the Python code. While I can sort of see and tolerate writing
`abs(Integer(-8)`  I am not convinced it is a good idea over writing `Expression(SymbolAbs, 8)`.

And I would not be surprised if down the line we come up with a better way of doing such things. 

So the question is whether in mathics Python code we would ever want to write `Integer(3) + String("a")` or the other thing.  

My take is that when the _need_ arises we can reconsider this. But I really doubt the right thing will ever be to put this in the base class. 

And right now clarity and simplicity is way more important that being able to handle hypothetical edge cases like the ones described above. 

Sure, inside Mathics, you should be able to do that. Inside Python code implementing Mathics, I'd need convincing though.

Reply:
> Nevertheless, from the Python side, maybe it also makes sense to define specialized behaviors for expressions like
> 
> ```
> Integer(1)+Integer(2)
> ```
> 
> or
> 
> ```
> Integer(3)+MachineReal(.12)
> ```

Ok. It is not clear that this is broken in the current code. None of the tests show it. If this isn't used right now in the code base let's leave it for later or feel free to at the mixin in the right place.

Reply:
My point is: in the Python side, usually you do not need to write ``abs(Integer(-8))`` but something like
```
def apply(expr, evaluation):
    '''F[expr_]'''
     ....
     new_expr = abs(expr)

```

in a way that ``element`` eventually is ``Integer(-8)`` or  ``Symbol("Pi")`` or ``Expression("F",37)``  or ``"Hello world"``. With this implementation, the last option would raise an exception. But OK, probably you would need to check that ``expr`` is something that makes sense to use as an argument of ``abs``.

Reply:
> > Nevertheless, from the Python side, maybe it also makes sense to define specialized behaviors for expressions like
> > ```
> > Integer(1)+Integer(2)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > or
> > ```
> > Integer(3)+MachineReal(.12)
> > ```
> 
> Ok. It is not clear that this is broken in the current code. None of the tests show it. If this isn't used right now in the code base let's leave it for later or feel free to add the mixin in the right place.

No, I do not mean that your changes break something. I'm OK with them.  What I am talking about is just to consider possible extensions, just to avoid repeating mistakes. I mean, maybe in the future we realize that makes sense to specialize __add__   for integers, in a way that instead of return an expression ``Expression("Plus",Integer(1),Integer(2))`` returns directly  ``Integer(3)`` (for example).

Reply:
Yeah, let's revisit this later.

Reply:
> 
> in a way that `element` eventually is `Integer(-8)` or `Symbol("Pi")` or `Expression("F",37)` or `"Hello world"`. With this implementation, the last option would raise an exception. But OK, probably you would need to check that `expr` is something that makes sense to use as an argument of `abs`.

Yes, I'd rather raise an exception here than east and execute on some mystery meat that might come back. (A friend of mine asked a Vietnamese person if they really ate dogs in Viet Nam; the response was "in the North only").